### PR TITLE
Current Windows Build

### PR DESCRIPTION
**Pull request by xTVaser**
Things left to do:
- Network stuff
- CI on windows

I added `FAKE_ISO_PATH` env-var instead of having it hardcoded.  Eventually we should just make a quick cross-platform filepath util function.

🙏 please don't fail linux build

### CODE REVIEW

File: game/kernel/kscheme.cpp [hunk 1]
```diff
@@ -318,24 +318,47 @@ u64 make_string_from_c(const char* c_str) {
...
  *
  * The implementation is to create a simple trampoline function which jumps to the C function.
+ *
+ *
  */
 Ptr<Function> make_function_from_c(void* func) {
   // allocate a function object on the global heap
   auto mem = Ptr<u8>(
-      alloc_heap_object(s7.offset + FIX_SYM_GLOBAL_HEAP, *(s7 + FIX_SYM_FUNCTION_TYPE), 0x40));
+      alloc_heap_object(s7.offset + FIX_SYM_GLOBAL_HEAP, *(s7 + FIX_SYM_FUNCTION_TYPE), 0x80));
```
Comment (xTVaser) [inline game/kernel/kscheme.cpp:327]: we did a lot of debugging in this file, not sure how much of this we want to keep

File: goalc/goos/Reader.cpp [hunk 2]
```diff
@@ -66,6 +66,10 @@ void TextStream::seek_past_whitespace_and_comments() {
   }
 }
 
+Reader::~Reader() {
```
Comment (xTVaser) [inline goalc/goos/Reader.cpp:69]: another file we did a lot of debugging in

**Review by water111 — COMMENTED** [review]
This looks great - getting a Windows build early on will be a huge help! This ended up being a lot more stuff than I thought.

There are a few small things that I think are broken on Linux right now due to some assembly changes that need to be windows specific.  Unfortunately we don't have any tests for them right now - I'm still working on the ability to start the runtime from within a gtest.  After that's working, this PR should be good to merge

File: test.sh [hunk 3]
```diff
@@ -4,4 +4,5 @@
 DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
 
 export NEXT_DIR=$DIR
+export FAKE_ISO_PATH=/game/fake_iso.txt
```
[the remark this answers is not in the document]
Comment (xTVaser) [reply test.sh:7]: Makes sense to me, I'll probably get some ideas once I start actually developing in this more.  Maybe some kind of configuration setup with some sane-defaults?

File: goalc/CMakeLists.txt [hunk 4]
```diff
@@ -1,9 +1,18 @@
...
+ELSE()
+  add_subdirectory(listener)
+ENDIF()
 add_subdirectory(emitter)
 
 add_executable(goalc main.cpp
         compiler/Compiler.cpp)
 
+IF (WIN32)
+  set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
[the remark this answers is not in the document]
Comment (xTVaser) [reply goalc/CMakeLists.txt:15]: Has something to do with making the shared linked libraries - https://stackoverflow.com/questions/40739061/error-lnk1104-cannot-open-file-debug-myprojectlib-lib/40743080#40743080

File: game/kernel/kprint.h [hunk 5]
```diff
@@ -65,25 +65,25 @@ void output_segment_load(const char* name, Ptr<u8> link_block, u32 flags);
 /*!
  * Print to the GOAL print buffer from C
  */
-void cprintf(const char* format, ...) __attribute__((format(printf, 1, 2)));
```
[the remark this answers is not in the document]
Comment (xTVaser) [reply game/kernel/kprint.h:68]: 👍 Can do, though I wondered if using a linter like https://clang.llvm.org/extra/clang-tidy/ might also catch such issues no matter the platform.

File: game/kernel/asm_funcs.asm [hunk 6]
```diff
@@ -32,10 +32,12 @@ _format:
   push rdi
 
   ; set the first argument register to the stack argument array
-  mov rdi, rsp
```
[the remark this answers is not in the document]
Comment (xTVaser) [reply game/kernel/asm_funcs.asm:35]: Hmm, difficult nasm can't really easily detect architecture apparently.  For now I'll just define two functions but probably something that can be improved going forward.

File: README.md [hunk 7]
```diff
@@ -163,7 +246,7 @@ The "runtime" will be a replacement for all of the C/C++ code of the original ga
   - [x] GOAL Linker
   - [ ] PS2-specific hardware initialization as required by Sony libraries
   - [x] GOAL "kheap" allocator
-  - [ ] Memory card interface
```
[the remark this answers is not in the document]
Comment (xTVaser) [reply README.md:166]: yup, there's some `interface` macro defined in gtest that maybe is only enabled on windows.  But I think it was causing problems for Doctor prompting a project-wide-find/replace.

File: game/kernel/asm_funcs.asm (hunk 6 above)
[the remark this answers is not in the document]
Comment (water111) [reply game/kernel/asm_funcs.asm:35]: That work for me!  I'm not too worried about this unless we start adding a bunch more assembly functions, but I hope to stick to just these two.

**Pull request merged**